# Convolutional Neural Network (CNN) implementation using TensorFlow

# CNN Components Explained

### Filter
In CNNs, a filter (also known as a kernel) is a small matrix used for convolution operations. Filters are applied to input data to extract features by performing element-wise multiplication and summation with the input data.

### Convolution Operation
The convolution operation involves applying a filter to an input image or feature map. This operation calculates the dot product between the filter and a small region of the input data, moving the filter across the entire input to produce a feature map.

### Convolution Layer
A convolutional layer consists of multiple filters that are convolved with the input data. Each filter learns to detect different features in the input. The output of each filter forms a separate feature map, which is then passed through an activation function to introduce non-linearity.

### Padding
Padding is the process of adding extra border pixels to the input data before applying convolution. Padding helps preserve spatial dimensions of the input and can be used to control the size of the output feature maps.

### Strides
Strides determine the step size with which the filter moves across the input data during convolution. A stride of 1 means the filter moves one pixel at a time, while a larger stride skips pixels, resulting in smaller output feature maps.

### Max Pooling
Max pooling is a downsampling operation commonly used in CNNs to reduce the spatial dimensions of the feature maps. It involves dividing the input into non-overlapping regions and selecting the maximum value from each region to form the pooled feature map. Max pooling helps in reducing computation and controlling overfitting.

### TensorFlow Implementation
In a TensorFlow implementation, you would typically define these components using TensorFlow's high-level API, such as `tf.keras.layers.Conv2D` for convolutional layers and `tf.keras.layers.MaxPooling2D` for max pooling layers. You would also specify parameters such as filter size, padding type, stride, and pooling size according to the requirements of your model.


In [1]:
import tensorflow as tf

In [2]:
# Define the input imahe shape(eg 28x28X1 for grayscale images)
input_shape = (28,28,1)

In [3]:
# Define the model

model = tf.keras.Sequential([
    # Convolutional layer with 32 filters , kernel size of 3x3 , RELU activation 
    tf.keras.layers.Conv2D(filters=32 , kernel_size=(3,3), activation='relu', input_shape=input_shape),
    
    # Adding padding to maintain spatial dimensions
    tf.keras.layers.ZeroPadding2D(padding=(1,1)),
    
    # Convolutional layer with 64 filters , kernel size of 3x3 , RELU activation 
    tf.keras.layers.Conv2D(filters=64 , kernel_size=(3,3), activation='relu'),
    
    # Max pooling layer with pool size of 2X2
    tf.keras.layers.MaxPool2D(pool_size=(2,2)),
    
    # Flatten layer to convert 2D features maps into 1D feature vector 
    tf.keras.layers.Flatten(),
    
    # Fully connected(dense) layer with units 128 and RELU activation
    
    tf.keras.layers.Dense(units=128 , activation='relu'),
    
    # Output layer with softmax activation for classification
    
    tf.keras.layers.Dense(units=10 , activation='softmax')
])

In [4]:
# Compile the model
model.compile(optimizer='adam',loss='sparse_categorical_crossentrophy' ,metrics=['accuracy'])

In [5]:
# Print the model summary
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 26, 26, 32)        320       
                                                                 
 zero_padding2d (ZeroPaddin  (None, 28, 28, 32)        0         
 g2D)                                                            
                                                                 
 conv2d_1 (Conv2D)           (None, 26, 26, 64)        18496     
                                                                 
 max_pooling2d (MaxPooling2  (None, 13, 13, 64)        0         
 D)                                                              
                                                                 
 flatten (Flatten)           (None, 10816)             0         
                                                                 
 dense (Dense)               (None, 128)               1

# Explanation of Shapes
- (None, 26, 26, 32): The first Conv2D layer outputs 32 feature maps of size 26x26 (due to the 3x3 kernel with no padding).
- (None, 28, 28, 32): The ZeroPadding2D layer adds padding, restoring the size to 28x28.
- (None, 26, 26, 64): The second Conv2D layer outputs 64 feature maps of size 26x26.
- (None, 13, 13, 64): The MaxPooling2D layer reduces the size by half to 13x13.
- (None, 10816): The Flatten layer reshapes the 13x13x64 feature maps into a 1D vector.
- (None, 128): The first Dense layer reduces the 10816 features to 128 units.
- (None, 10): The output Dense layer reduces the 128 units to 10 units (for classification).

## Why Not Use 1D or 3D?
##### 1D Layers (Conv1D, MaxPooling1D): Used for sequential data, like time series or audio signals, where the data has a single spatial dimension.
###### 3D Layers (Conv3D, MaxPooling3D): Used for volumetric data, like video or 3D medical imaging, where the data has three spatial dimensions (height, width, and depth).